[Materials] Add rendering support for hosted materials

pxlcoder opened the pull request:
#### 99b9a154a6f7e44e8a17c81b12919b8bf76fa6ce
<pre>
[Materials] Add rendering support for hosted materials
<a href="https://bugs.webkit.org/show_bug.cgi?id=287421">https://bugs.webkit.org/show_bug.cgi?id=287421</a>
<a href="https://rdar.apple.com/144551395">rdar://144551395</a>

Reviewed by Richard Robinson.

Support rendering materials implemented in SwiftUI. These materials can be used
via `-apple-visual-effect`.

When a &quot;hosted material&quot; is used, a structural layer is created to encapsulate
the content layer. In the UI process, the structural layer is realized using
SwiftUI. On iOS, a `_UIHostingView` is created, and on macOS a `_CALayerView`
is created. A `materialEffect` is then applied to their respective root views.
WebKit&apos;s own layers are parented underneath the SwiftUI-created views/layers.

* Source/WebCore/platform/cocoa/AppleVisualEffect.cpp:
(WebCore::appleVisualEffectIsHostedMaterial):
* Source/WebCore/platform/cocoa/AppleVisualEffect.h:

Introduce a `borderRect` property to `AppleVisualEffectData`. This is necessary
as the shape of the material needs to be applied to the structural layer via
the `materialEffect`, and cannot be set via traditional `CALayer` properties.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):

Ensure a structural layer is created for hosted materials, using `-apple-visual-effect.`

(WebCore::GraphicsLayerCA::updateNames):
(WebCore::GraphicsLayerCA::updateAppleVisualEffectData):
(WebCore::GraphicsLayerCA::ensureStructuralLayer):

Creating a structural layer for material hosting takes precedence over
preserves-3D and replica flattening. Those features are not supported with
hosted materials, due to a (current) lack of use case.

(WebCore::GraphicsLayerCA::structuralLayerPurpose const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::PlatformCALayerCocoa):

Specify a plain layer here, as the feature is unsupported when UI-side
compositing is disabled.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAppleVisualEffect):

Specify the `borderRect` on the `AppleVisualEffectData` so that the
material effect can set a corner radius.

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.h: Added.

Expose utilities from Swift to Obj-C to work with material effects.

* Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift: Added.
(MaterialHostingContentViewWrapper.makeUIView(_:)):

A `UIViewRepresentable` is necessary since WebKit&apos;s own views need to be
parented underneath the SwiftUI view with the material effect applied.

(MaterialHostingContentViewWrapper.updateUIView(_:context:)):
(MaterialHostingProvider.view(for:)):

Add `MaterialHostingProvider` as a protocol to vend layer-backed or
view-backed material effects.

(LayerBackedMaterialHostingProvider.view(for:)):

The layer-backed material provider uses a `_CALayerView`. WebKit&apos;s own layers
are parented underneath the created layer.

(ViewBackedMaterialHostingProvider.view(for:)):

The view-backed material uses a `UIViewRepresentable`. WebKit&apos;s own views are
parented underneath the `UIViewRepresentable`.

(MaterialHostingView.body):

Share code to either return a plain SwiftUI view (from the material hosting provider)
or one with the material effect applied. `MaterialHostingView` is specialized on
`MaterialHostingProvider` to allow for code reuse.

(CALayer.materialHostingContentLayer):

Add an extension to `CALayer` to make it easy to get/set the content layer that
will parent WebKit&apos;s own layers.

(WKMaterialHostingSupport.createHostingLayer):

Create two layers. A `CAHostingLayer` to encapsulate a SwiftUI view with a
material effect, and a regular `CALayer` which will parent WebKit&apos;s own layers.

The `contentLayer` is stored using a key on the `CAHostingLayer` so that it
may be retrieved later to parent WebKit&apos;s own layers.

(WKMaterialHostingSupport.updateHostingLayer(_:cornerRadius:)):

Apply the material effect. This involves recreating the `_CALayerView`, so the
content layer must be re-parented under the new view/layer.

(WKMaterialHostingSupport.contentLayer(forMaterialHostingLayer:)):

Expose the `contentLayer` in order to add WebKit&apos;s layers as sublayers.

(WKMaterialHostingSupport.createHostingView(_:)):

Create a `_UIHostingView` which encapsulates a content view specified by the
remote layer tree.

(WKMaterialHostingSupport.updateHostingView(_:contentView:cornerRadius:)):

Apply the material effect. This involves recreating the wrapper view,
and content view must be re-parented under the new view.

* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::updateAppleVisualEffect):

Refactor the `AppleVisualEffectChanged` update into a method.

Use the `MaterialHostingSupport` utilities to apply the new effects.

(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyHierarchyUpdates):

Access the content layer / content view on the structural layer / view so that
WebKit&apos;s own layers are parented underneath a view with material effects applied.

If this were not done, the material effect would be applied to an (empty) sibling
layer, which would be incorrect.

(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToUIView):

Since `WKMaterialHostingView` has a intermediate &quot;hosting view&quot;, its size must
be updated along with the `WKMaterialHostingView` itself. Without this, the
`_UIHostingView` would have a zero frame, and SwiftUI would not size the `rootView`.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(-[WKMaterialHostingView init]):

`WKMaterialHostingView` has a hosting view (which is the `_UIHostingView`) and
a content view (which is a subview of the hosting view and serves to parent
WebKit&apos;s own views).

`WKMaterialHostingView` is used rather than simply returning a `_UIHostingView`
when creating the structural node for two reasons:
1. To easily give access the content view.
2. To subclass `WKCompositingView`.

(-[WKMaterialHostingView contentView]):
(-[WKMaterialHostingView updateHostingSize:]):
(-[WKMaterialHostingView updateCornerRadius:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/290280@main">https://commits.webkit.org/290280@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e23267c8c8bb8be0d61b9bf6bbb16ea71f2c8a7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89456 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94447 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40223 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91508 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9370 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17262 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68912 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26573 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92457 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7193 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81183 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49276 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6928 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35565 "Found 3 new test failures: fonts/monospace.html fonts/sans-serif.html fonts/serif.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39329 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77286 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36566 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96276 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16637 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12229 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77784 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76983 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77098 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21524 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20103 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9788 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14039 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16651 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21960 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16391 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19842 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18173 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+/*
```
we generally use // for copyright comments in Swift files

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+        contentView
+    }
+
+    func updateUIView(_ uiView: UIView, context: Context) { }
```
```suggestion
    func updateUIView(_ uiView: UIView, context: Context) { 
    }
```

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+#endif
+
+@objc(WKMaterialHostingSupport)
+class MaterialHostingSupport: NSObject {
```
this should be `@objc @implementation extension WKMaterialHostingSupport`

pxlcoder replied:
`error: cannot find type 'WKMaterialHostingSupport' in scope`

pxlcoder replied:
What I have also matches other Swift usage in WebKit...

pxlcoder replied:
Done!

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+            layer.addSublayer(contentLayer)
+        }))
+
+        hostingLayer.setValue(contentLayer, forKeyPath: materialHostingContentLayerKey)
```
this pains me, but I get it

pxlcoder replied:
Slightly refactored.

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+    class func createHostingLayer() -> CALayer {
+        let contentLayer = CALayer()
+
+        let hostingLayer = CAHostingLayer(rootView: AnyView(_CALayerView(type: CALayer.self) { layer in
```
this SwiftUI view should be extracted into its own type

pxlcoder replied:
Don't really understand this comment, let's talk tomorrow!

pxlcoder replied:
Done!

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+    class func createHostingLayer() -> CALayer {
+        let contentLayer = CALayer()
+
+        let hostingLayer = CAHostingLayer(rootView: AnyView(_CALayerView(type: CALayer.self) { layer in
```
why does this need the AnyView again?

pxlcoder replied:
Because `.materialEffect` returns `some View`, and the `rootView` is updated later.

pxlcoder replied:
Got rid of `AnyView`!

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+    @objc(updateHostingLayer:cornerRadius:)
+    class func updateHostingLayer(_ layer: CALayer, cornerRadius: CGFloat) {
+        guard let hostingLayer = layer as? CAHostingLayer<AnyView> else {
+            return;
```
maybe add an assertionFailure here (which is debug only)

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+            return;
+        }
+
+        guard let contentLayer = hostingLayer.value(forKeyPath: materialHostingContentLayerKey) as? CALayer else {
```
I think maybe this can be refactored into a bit more type safe extension function

pxlcoder replied:
Will need your help with that.

pxlcoder replied:
Done!

rr-codes commented on Source/WebKit/Platform/cocoa/MaterialHostingSupport.swift:
```diff
@@ -0,0 +1,109 @@
+
+#if canImport(UIKit)
+
+    @MainActor @objc(createHostingView:)
```
MainActor shouldn't be needed

pxlcoder replied:
It is needed.

```
error: main actor-isolated property 'rootView' can not be mutated from a nonisolated context
```

pxlcoder replied:
Removed after specifying `NS_SWIFT_UI_ACTOR` in the header.

rr-codes commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
```diff
@@ -392,6 +396,46 @@ + (Class)layerClass
 
+#if HAVE(MATERIAL_HOSTING)
+
+@implementation WKMaterialHostingView {
```
why can't this be in Swift again?

pxlcoder replied:
So that it can subclass `WKCompositingView` and conform to `WKContentControlled`.

rr-codes commented on Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
```diff
@@ -0,0 +1,159 @@
+#endif
+
+struct WKMaterialHostingLayerView: View {
+    let contentLayer: CALayer
```
these properties should be private

pxlcoder replied:
Done.

rr-codes commented on Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
```diff
@@ -0,0 +1,159 @@
+@_spi(Private) @_spi(ForAppKitOnly) internal import SwiftUI
+#endif
+
+struct WKMaterialHostingLayerView: View {
```
this struct should be private (its only used in this file, right?)

pxlcoder replied:
Done.

rr-codes commented on Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
```diff
@@ -0,0 +1,159 @@
+
+#endif
+
+extension CALayer {
```
this can also be private

pxlcoder replied:
Done.

rr-codes commented on Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
```diff
@@ -0,0 +1,159 @@
+#if canImport(UIKit)
+
+    class func createHostingView(_ contentView: UIView) -> UIView {
+        return _UIHostingView(rootView: WKMaterialHostingRootView(contentView: contentView))
```
return isn't needed

pxlcoder replied:
Removed.

pxlcoder committed:
[Materials] Add rendering support for hosted materials
https://bugs.webkit.org/show_bug.cgi?id=287421
rdar://144551395

Reviewed by Richard Robinson.

Support rendering materials implemented in SwiftUI. These materials can be used
via `-apple-visual-effect`.

When a "hosted material" is used, a structural layer is created to encapsulate
the content layer. In the UI process, the structural layer is realized using
SwiftUI. On iOS, a `_UIHostingView` is created, and on macOS a `_CALayerView`
is created. A `materialEffect` is then applied to their respective root views.
WebKit's own layers are parented underneath the SwiftUI-created views/layers.

* Source/WebCore/platform/cocoa/AppleVisualEffect.cpp:
(WebCore::appleVisualEffectIsHostedMaterial):
* Source/WebCore/platform/cocoa/AppleVisualEffect.h:

Introduce a `borderRect` property to `AppleVisualEffectData`. This is necessary
as the shape of the material needs to be applied to the structural layer via
the `materialEffect`, and cannot be set via traditional `CALayer` properties.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):

Ensure a structural layer is created for hosted materials, using `-apple-visual-effect.`

(WebCore::GraphicsLayerCA::updateNames):
(WebCore::GraphicsLayerCA::updateAppleVisualEffectData):
(WebCore::GraphicsLayerCA::ensureStructuralLayer):

Creating a structural layer for material hosting takes precedence over
preserves-3D and replica flattening. Those features are not supported with
hosted materials, due to a (current) lack of use case.

(WebCore::GraphicsLayerCA::structuralLayerPurpose const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::PlatformCALayerCocoa):

Specify a plain layer here, as the feature is unsupported when UI-side
compositing is disabled.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAppleVisualEffect):

Specify the `borderRect` on the `AppleVisualEffectData` so that the
material effect can set a corner radius.

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.h: Added.

Expose utilities from Swift to Obj-C to work with material effects.

* Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift: Added.
(MaterialHostingContentViewWrapper.makeUIView(_:)):

A `UIViewRepresentable` is necessary since WebKit's own views need to be
parented underneath the SwiftUI view with the material effect applied.

(MaterialHostingContentViewWrapper.updateUIView(_:context:)):
(MaterialHostingProvider.view(for:)):

Add `MaterialHostingProvider` as a protocol to vend layer-backed or
view-backed material effects.

(LayerBackedMaterialHostingProvider.view(for:)):

The layer-backed material provider uses a `_CALayerView`. WebKit's own layers
are parented underneath the created layer.

(ViewBackedMaterialHostingProvider.view(for:)):

The view-backed material uses a `UIViewRepresentable`. WebKit's own views are
parented underneath the `UIViewRepresentable`.

(MaterialHostingView.body):

Share code to either return a plain SwiftUI view (from the material hosting provider)
or one with the material effect applied. `MaterialHostingView` is specialized on
`MaterialHostingProvider` to allow for code reuse.

(CALayer.materialHostingContentLayer):

Add an extension to `CALayer` to make it easy to get/set the content layer that
will parent WebKit's own layers.

(WKMaterialHostingSupport.createHostingLayer):

Create two layers. A `CAHostingLayer` to encapsulate a SwiftUI view with a
material effect, and a regular `CALayer` which will parent WebKit's own layers.

The `contentLayer` is stored using a key on the `CAHostingLayer` so that it
may be retrieved later to parent WebKit's own layers.

(WKMaterialHostingSupport.updateHostingLayer(_:cornerRadius:)):

Apply the material effect. This involves recreating the `_CALayerView`, so the
content layer must be re-parented under the new view/layer.

(WKMaterialHostingSupport.contentLayer(forMaterialHostingLayer:)):

Expose the `contentLayer` in order to add WebKit's layers as sublayers.

(WKMaterialHostingSupport.createHostingView(_:)):

Create a `_UIHostingView` which encapsulates a content view specified by the
remote layer tree.

(WKMaterialHostingSupport.updateHostingView(_:contentView:cornerRadius:)):

Apply the material effect. This involves recreating the wrapper view,
and content view must be re-parented under the new view.

* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::updateAppleVisualEffect):

Refactor the `AppleVisualEffectChanged` update into a method.

Use the `MaterialHostingSupport` utilities to apply the new effects.

(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyHierarchyUpdates):

Access the content layer / content view on the structural layer / view so that
WebKit's own layers are parented underneath a view with material effects applied.

If this were not done, the material effect would be applied to an (empty) sibling
layer, which would be incorrect.

(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToUIView):

Since `WKMaterialHostingView` has a intermediate "hosting view", its size must
be updated along with the `WKMaterialHostingView` itself. Without this, the
`_UIHostingView` would have a zero frame, and SwiftUI would not size the `rootView`.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(-[WKMaterialHostingView init]):

`WKMaterialHostingView` has a hosting view (which is the `_UIHostingView`) and
a content view (which is a subview of the hosting view and serves to parent
WebKit's own views).

`WKMaterialHostingView` is used rather than simply returning a `_UIHostingView`
when creating the structural node for two reasons:
1. To easily give access the content view.
2. To subclass `WKCompositingView`.

(-[WKMaterialHostingView contentView]):
(-[WKMaterialHostingView updateHostingSize:]):
(-[WKMaterialHostingView updateCornerRadius:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/290280@main

axiixc commented on Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
```diff
@@ -0,0 +1,162 @@
+        let view = P.view(for: content)
+
+        if useMaterialEffect {
+            view.materialEffect(Material._hostedBlurMaterial, in: .rect(cornerRadius: cornerRadius))
```
This will probably be fine in the short term, but using `if` this way is a bit of an anti-pattern due to invalidating the view’s identity between branches. If `view` had internal state this could result in it flip-flopping or being lost when the condition changes. Without the definition for `materialEffect(_:in:)` it’s hard to say, but generally SwiftUI modifiers should provide you an inert form so that you can sink those conditions into the arguments, while “structurally” the modifiers remain unconditional.

I also kinda wonder if a `ViewModifier` might be more correct here, just given how the structure of this looks.

pxlcoder replied:
Thanks for this, and the other feedback. I've filed https://bugs.webkit.org/show_bug.cgi?id=287735 to track addressing the feedback.

axiixc commented on Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
```diff
@@ -0,0 +1,162 @@
+
+    var materialHostingContentLayer: CALayer? {
+        get {
+            guard let contentLayer = self.value(forKeyPath: Self.materialHostingContentLayerKey) as? CALayer else {
```
Nit: You could just return this directly since it’s already optional.
```swift
get { value(forKeyPath: Self.materialHostingContentLayerKey) as? CALayer? }
```

axiixc commented on Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
```diff
@@ -0,0 +1,162 @@
+#if canImport(UIKit)
+
+private struct MaterialHostingContentViewWrapper: UIViewRepresentable {
+    let contentView: UIView
```
There is a subtle gotcha here. Since you’re returning this directly from `makeUIView()` you could end up with stale/untracked updates in the future. This is a bit contrived, but basically imagine a setup like this…
```swift
struct BadBehavior : View {
    var body: some View {
        MaterialHostingContentViewWrapper(contentView: Bool.random() ? ViewA.shared : View.B.shared)
    }
}
```
Here, the identity of the representable is stable, but when the body of `BadBehavior` is recomputed it receives a random `contentView`. Which view will show up basically gets fixed on the first render though, since `makeUIView()` will only be called once.

The more generalized fix would be wrapping the result view, so basically returning some dummy `UIView()` instance as `makeUIView()`’s result, and then adding/removing `contentView` in `updateUIView()` instead[^1]. That is a lot more overhead and bookkeeping (and could also have implications on sizing/layout as well), so not really suggesting changes at this time. Especially because in a tightly controlled scenario like this you’re probably safe, but I did want to bring this up as a warning in case this pattern gets copied going forward in more complex views in the future.

[^1]: And SwiftUI will _already_ wrap your returned view in its own to isolate it from the private view hierarchy somewhat, so it’s extra-extra overhead to boot!
